File: 494-TargetSum/494-TargetSum.java
/*

494. Target Sum
Medium

You are given a list of non-negative integers, a1, a2, ..., an, and a target, S. Now you have 2 symbols + and -. For each integer, you should choose one from + and - as its new symbol.

Find out how many ways to assign symbols to make sum of integers equal to target S.

Example 1:

Input: nums is [1, 1, 1, 1, 1], S is 3. 
Output: 5
Explanation: 

-1+1+1+1+1 = 3
+1-1+1+1+1 = 3
+1+1-1+1+1 = 3
+1+1+1-1+1 = 3
+1+1+1+1-1 = 3

There are 5 ways to assign symbols to make the sum of nums be target 3.

Note:

    The length of the given array is positive and will not exceed 20.
    The sum of elements in the given array will not exceed 1000.
    Your output answer is guaranteed to be fitted in a 32-bit integer.

*/

class Solution {
    int count = 0;
    public int findTargetSumWays(int[] nums, int S) {
        dfs(nums, S, 0, 0);
        return count;
    }
    private void dfs(int[] nums, int S, int curr, int idx){
        if(idx == nums.length){
            if(curr == S)
                count++;
             return;
        }
        dfs(nums, S, curr + nums[idx], idx + 1);
        dfs(nums, S, curr - nums[idx], idx + 1);
    }
}

class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        return dfs(nums, S, 0, 0);
    }
    private int dfs(int[] nums, int S, int curr, int idx){
        if(idx == nums.length){
            if(curr == S)
               return 1;
             else 
                 return 0;
        }
        return dfs(nums, S, curr + nums[idx], idx + 1) + 
        dfs(nums, S, curr - nums[idx], idx + 1);
    }
}

class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        //DP with memoization
        int sum = 0;
        for (int num : nums) 
            sum += num;
        Integer[][] cache = new Integer[nums.length + 1][2 * sum + 1];
        
        return dfs(nums, sum, S + sum, 0, cache);
    }
    
    private int dfs(int[] nums, int cur, int target, int idx, Integer[][] cache) {
        if (idx == nums.length) {
            return cur == target ? 1 : 0;
        }
        if (cache[idx][cur] != null) {
            return cache[idx][cur];
        }
        
        int res = 0;
        res = dfs(nums, cur + nums[idx], target, idx + 1, cache)
            + dfs(nums, cur - nums[idx], target, idx + 1, cache);
        
        cache[idx][cur] = res;
        return res;
    }
}

File: 494-TargetSum/494-TargetSum.py
"""

494. Target Sum

You are given an integer array nums and an integer target.

You want to build an expression out of nums by adding one of the symbols '+' and '-' before each integer in nums and then concatenate all the integers.

    For example, if nums = [2, 1], you can add a '+' before 2 and a '-' before 1 and concatenate them to build the expression "+2-1".

Return the number of different expressions that you can build, which evaluates to target.

Example 1:

Input: nums = [1,1,1,1,1], target = 3
Output: 5
Explanation: There are 5 ways to assign symbols to make the sum of nums be target 3.
-1 + 1 + 1 + 1 + 1 = 3
+1 - 1 + 1 + 1 + 1 = 3
+1 + 1 - 1 + 1 + 1 = 3
+1 + 1 + 1 - 1 + 1 = 3
+1 + 1 + 1 + 1 - 1 = 3

Example 2:

Input: nums = [1], target = 1
Output: 1

Constraints:

    1 <= nums.length <= 20
    0 <= nums[i] <= 1000
    0 <= sum(nums[i]) <= 1000
    -1000 <= target <= 1000

"""

class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        def dfs(curr, idx):
            key = (idx, curr)
            if key in cache: return cache[key]
            if idx == len(nums):
                return 1 if curr == target else 0
            else:
                add = dfs(curr + nums[idx], idx + 1)
                sub = dfs(curr - nums[idx], idx + 1)
                cache[key] = add + sub
                return cache[key]
        cache = defaultdict(int)
        return dfs(0, 0)
